subsys/ipc/ipc_service/backends/ipc_icmsg_with_buf.c

```diff
@@ -0,0 +1,1222 @@
+ *
+ * @return	zero or ICMsg send error
+ */
+static int send_release(struct backend_data *dev_data, const uint8_t *buffer,
+			uint8_t msg_type)
+{
+	const struct icmsg_with_buf_config *conf = dev_data->conf;
+	int rx_block_index;
+
+	rx_block_index = buffer_to_index_validate(&conf->rx, buffer, NULL);
+	if (rx_block_index < 0) {
+		return rx_block_index;
+	}
+
+	return icmsg_send_wrapper(dev_data, msg_type, rx_block_index);
+}
+
+/**
+ * Send data contained in specified block. It will adjust data size and flush cache
+ * if necessary. If sending failed, allocated blocks will be released.
+ *
+ * @param[in] tx_block_index	Index of first block containing data, it is not validated,
+ *				so caller is responsible for passing only valid index.
+ * @param[in] size		Actual size of the data, can be smaller than allocated,
+ *				but it cannot change number of required blocks.
+ * @param[in] remote_addr	Remote endpoints address
+ *
+ * @return			O or negative error code
+ */
+static int send_block(struct backend_data *dev_data, size_t tx_block_index, size_t size,
+		      uint8_t remote_addr)
+{
+	struct block_header *block;
+	int r;
+
+	block = block_from_index(&dev_data->conf->tx, tx_block_index);
+
+	block->size = size;
+	__sync_synchronize();
+	sys_cache_data_flush_range(block, size + BLOCK_HEADER_SIZE);
+
+	r = icmsg_send_wrapper(dev_data, remote_addr, tx_block_index);
+
+	if (r < 0) {
+		release_tx_blocks(dev_data, tx_block_index, size, -1);
+	}
+
+	return r;
+}
+
+/**
+ * Find endpoint that was registered with name that matches name
+ * contained in the endpoint bound message received from remote. This function
+ * must be called when mutex is locked.
+ *
+ * @param[in] name	Endpoint name, it must be in a received block.
+ *
+ * @return	Found endpoint data or NULL if not found
+ */
+static struct ept_data *find_ept_by_name(struct backend_data *dev_data, const char *name)
+{
+	const struct channel_config *rx_conf = &dev_data->conf->rx;
+	const char *buffer_end = (const char *)rx_conf->blocks_ptr +
+				 rx_conf->block_count * rx_conf->block_size;
+	size_t name_size;
+	size_t i;
+
+	/* Requested name is in shared memory, so we have to assume that it
+	 * can be corrupted. Extra care must be taken to avoid out of
+	 * bounds reads.
+	 */
+	name_size = strnlen(name, buffer_end - name - 1) + 1;
+
+	for (i = 0; i < dev_data->ept_count; i++) {
+		if (strncmp(dev_data->ept[i].cfg->name, name, name_size) == 0) {
+			return &dev_data->ept[i];
+		}
+	}
+	return NULL;
+}
+
+/**
+ * Find registered endpoint that matches given "bound endpoint" message. When found,
+ * remote endpoint address is saved for future use and "release bound endpoint" message
+ * is send. This function must be called when mutex is locked.
+ *
+ * @param[in] rx_block_index	Block containing the "bound endpoint" message.
+ *
+ * @retval 0	match not found
+ * @retval 1	match found and processing was successful
+ * @retval < 0	error occurred during processing
```

Comment:
`retval < 0` ?

Reply:
Ok, it is not valid doxygen syntax. I will fix that.